source/blog/2016-04-01-deadlock-investigation.html.md

```diff
@@ -0,0 +1,191 @@
+what services the user is allowed to see.  The rest api is sitting behind apache
+running rails 5 using puma as the web server.  Note, puma spins up threads to
+process requests.
+
+With Matthew Draper's help, I started to investigate it.
+
+The first step was to look in our Rails production.log:
+
+```
+[----] I, [2016-03-31T18:08:12.327408 #7579:12a1074]  INFO -- : Started GET "/api/auth?requester_type=ui" for 127.0.0.1 at 2016-03-31 18:08:12 -0400
+[----] I, [2016-03-31T18:08:13.767557 #7579:12a1074]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:13.767673 #7579:12a1074]  INFO -- :   Parameters: {"requester_type"=>"ui", "collection"=>"auth"}
+[----] I, [2016-03-31T18:08:13.768477 #7579:12a1074]  INFO -- : Fork detected, re-connecting child process...
+[----] I, [2016-03-31T18:08:13.966210 #7579:12a1074]  INFO -- : Completed 200 OK in 198ms (Views: 0.6ms | ActiveRecord: 0.0ms)
+[----] I, [2016-03-31T18:08:13.974621 #7579:12a1074]  INFO -- : Started GET "/api?attributes=authorization" for 127.0.0.1 at 2016-03-31 18:08:13 -0400
+[----] I, [2016-03-31T18:08:13.992612 #7579:12a1074]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:13.992683 #7579:12a1074]  INFO -- :   Parameters: {"attributes"=>"authorization"}
+[----] I, [2016-03-31T18:08:15.895767 #7579:12a1074]  INFO -- : Completed 200 OK in 1903ms (Views: 0.4ms | ActiveRecord: 0.0ms)
+[----] I, [2016-03-31T18:08:15.916145 #7579:12a1074]  INFO -- : Started GET "/api/services?filter[]=service_id=nil" for 127.0.0.1 at 2016-03-31 18:08:15 -0400
+[----] I, [2016-03-31T18:08:15.943982 #7579:12a1074]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:15.944103 #7579:12a1074]  INFO -- :   Parameters: {"filter"=>["service_id=nil"], "collection"=>"services"}
+[----] I, [2016-03-31T18:08:15.989375 #7579:3e35e24]  INFO -- : Started GET "/api/services?filter[]=retired=false" for 127.0.0.1 at 2016-03-31 18:08:15 -0400
+[----] I, [2016-03-31T18:08:16.010779 #7579:3e35c08]  INFO -- : Started GET "/api/services?filter[]=retired=true" for 127.0.0.1 at 2016-03-31 18:08:16 -0400
+[----] I, [2016-03-31T18:08:16.029485 #7579:3e35e24]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:16.029638 #7579:3e35e24]  INFO -- :   Parameters: {"filter"=>["retired=false"], "collection"=>"services"}
+[----] I, [2016-03-31T18:08:16.054986 #7579:3e35c08]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:16.055079 #7579:3e35c08]  INFO -- :   Parameters: {"filter"=>["retired=true"], "collection"=>"services"}
+```
+
+What is interesting is we see threads **3e35e24** and **3e35c08** in process
+**7579** are both processing `ApiController#show` but never complete.
+
+They seem to be victims of a deadlock.
+
+So, let's see what's happening...
+
+```
+kill -19 7452  # SIGSTOP 7579's parent where it is so it doesn't kill our process
```

Comment:
I need to better highlight this insight by @matthewd.  I've seen other places show gdb with ruby but stopping the parent process is a great idea to avoid an external process from messing with our process.